test/unit-tests.c

```diff
@@ -1202,11 +1203,157 @@ test_migration_state_transitions(void **state __attribute__ ((unused)))
  * and provide a function to execute before and after each unit test.
  */
 static int
-setup(void **state __attribute__((unused))) {
+setup(void **state __attribute__((unused)))
+{
     unpatch_all();
     return 0;
 }
 
+static struct test_setup_migr_reg_dat {
+    vfu_ctx_t *v;
+    size_t rs; /* migration registers size */
+    size_t ds; /* migration data size */
+    size_t s; /* migration region size*/
+    const vfu_migration_callbacks_t c;
+} migr_reg_data = {
+    .c = {
+        .version = VFU_MIGR_CALLBACKS_VERS,
+        .transition = (void*)0x1,
+        .get_pending_bytes = (void*)0x2,
+        .prepare_data = (void*)0x3,
+        .read_data = (void*)0x4,
+        .write_data = (void*)0x5,
+        .data_written = (void*)0x6
+    }
+};
+
+static int
+setup_test_setup_migration_region(void **state) {
+    struct test_setup_migr_reg_dat *p = &migr_reg_data;
+    p->v = vfu_create_ctx(VFU_TRANS_SOCK, "test", 0, NULL,
+        VFU_DEV_TYPE_PCI);
+    if (p->v == NULL) {
+        return -1;
+    }
+    p->rs = ROUND_UP(sizeof(struct vfio_device_migration_info), sysconf(_SC_PAGE_SIZE));
+    p->ds = sysconf(_SC_PAGE_SIZE);
+    p->s = p->rs + p->ds;
+    *state = p;
+    return setup(state);
+}
+
+static vfu_ctx_t *
+get_vfu_ctx(void **state)
+{
+    return (*((struct test_setup_migr_reg_dat**)(state)))->v;
+}
+
+static int
+teardown_test_setup_migration_retion(void **state)
+{
+    struct test_setup_migr_reg_dat *p = *state;
+    vfu_destroy_ctx(p->v);
+    return 0;
+}
+
+static void
+test_setup_migration_region_too_small(void **state)
+{
+    vfu_ctx_t *v = get_vfu_ctx(state);
+    int r = vfu_setup_region(v, VFU_PCI_DEV_MIGR_REGION_IDX,
+        sizeof(struct vfio_device_migration_info) - 1, NULL,
+        VFU_REGION_FLAG_READ | VFU_REGION_FLAG_WRITE, NULL, 0, -1);
+    assert_int_equal(-1, r);
+    assert_int_equal(EINVAL, errno);
+}
+
+static void
+test_setup_migration_region_size_ok(void **state)
+{
+    vfu_ctx_t *v = get_vfu_ctx(state);
+    int r = vfu_setup_region(v, VFU_PCI_DEV_MIGR_REGION_IDX,
+        sizeof(struct vfio_device_migration_info), NULL,
+        VFU_REGION_FLAG_READ | VFU_REGION_FLAG_WRITE, NULL, 0, -1);
+    assert_int_equal(0, r);
+}
+
+static void
```

Comment:
I'm wondering if we should handle this ourselves. Not sure

Reply:
Do you mean making the migration region mappable without telling the user?

Reply:
So for non-migration regions, if an fd is provided, but no mmap areas, we map the whole region.
I'm saying that for the migration region, if an fd is provided , but no mmap areas, could we map the whole region, excluding the first page?

Reply:
We can't assume that the space after the migration registers is simply padding, the user might setup the migration region without calling `vfu_get_migr_register_area_size` and putting data right after the migration registers.

Reply:
I'm not sure why we need to support that (but maybe we do?). Isn't it simpler to have a simple "page size" requirement?

Reply:
We don't necessarily need to support that, but seems that we're making it more complicated (a) by introducing a requirement, and (b) by changing the semantics of the how the function works if mmap areas is NULL but only if this is the migration region. I think it will be harder to get it right this way.

Reply:
ok, but I think we should always insist on a page-sized buffer, regardless.

Reply:
> ok, but I think we should always insist on a page-sized buffer, regardless.

We sure can, but if we're introducing a restriction we should explain why?

I think if we want to make the API easier we can explain that the simplest way to make the migration region mappable is for the client to use `vfu_get_migr_register_area_size` in order to allocate space for the migration registers and then

Reply:
Right, I think we're agreeing: we should say that regardless of whether you're using the callbacks, and regardless of if you're providing an fd, you need to reserve a non-mappable area of size vfu_get_migr_register_area_size().

Reply:
I re-read the PR and it's currently implemented the way you suggest:
```
The migration region must contain at the beginning the migration
registers (struct vfio_device_migration_info defined in <linux/vfio.h>) and
the remaining part of the region can be arbitrarily used by the device
implementation. The region provided must have at least
vfu_get_migr_register_area_size() bytes available at the start of the region
(this size is guaranteed to be page-aligned).
```
So going back to your original comment in this discussion, we're deciding _not_ to special case the `mmap_areas == NULL` case for the migration region, right?

Reply:
We won't special case.

But it's not implemented as you suggest, yet:

```
   return ranges_intersect(0, sizeof(struct vfio_device_migration_info),
                            (size_t)iov->iov_base, (size_t)iov->iov_len);
```

```
    if (region_idx == VFU_PCI_DEV_MIGR_REGION_IDX &&
        size < sizeof(struct vfio_device_migration_info)) {
```

At least both of these are using the struct size not the vfu_get_migr_register_area_size()

Reply:
Right, this needs fixing.